### DISCUSSION TITLE
How to run a recipe only if a particular library is present AND not run another recipe when that dependency is present

### DISCUSSION THREAD

**Discussion by gjd6640** [openrewrite/rewrite#3569] [Q&A] [answered]
I have a scenario where I want to run a recipe that'll upgrade to Spring 4 when a particular library is present in a Maven project. Otherwise, I want a different recipe to run that'll upgrade to Spring 5. Is there a way to automate that decision via recipe configs in my rewrite.yml? For example, if "onlyIfUsing" supported version number matching on the major version that might enable me to make the correct recipe run based on whether the problematic library is present.

I see that [org.openrewrite.maven.search.DoesNotIncludeDependency](https://docs.openrewrite.org/recipes/maven/search/doesnotincludedependency) exists but cannot find any discussion or end-user usage examples. Not sure if this is what it is for or not...

**Comment by timtebeek** [comment, marked as answer]
Hi! We [added Preconditions in Rewrite 8](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices#use-preconditions) to allow you to express conditions that should match before your recipe runs. In your case you can likely develop two custom recipes:
- one with a precondition that checks the library is present,
- and another with a precondition that checks the library is not present.

Then, for each of those recipes, you can either [chain visitors](https://docs.openrewrite.org/concepts-explanations/visitors#chaining-visitors), or [link recipes together](https://docs.openrewrite.org/concepts-explanations/recipes#linking-recipes-together) to run the additional existing recipes that you want to invoke.

That way you have relatively straightforward individual recipes that each match in specific conditions, and only then make their changes, instead of a single recipe that always matches and attempts changes, with more complicated decision logic that's harder to test and reason about.

Hope that helps!

**Reply by gjd6640** [reply]
Thank you for the prompt and clear response. I was hoping to find that there's a way to apply a precondition from a YAML config file since we're new to the product and are exploring using the product entirely via YAML configs right now. Per the [YAML format reference](https://docs.openrewrite.org/reference/yaml-format-reference) documentation page it looks like that's not supported at this time:

> Declarative YAML recipes do not support preconditions (formerly named applicability tests) as of [OpenRewrite 8](https://docs.openrewrite.org/changelog/8-1-2-release#applicability-tests-have-been-replaced).

Thanks again. In the meantime, we'll plan to just maintain several recipes in our YAML config and we'll externally determine which recipes make sense to run.

**Reply by timtebeek** [reply]
Glad to hear you've found a form that works for you right now. Wishing you all the best and let us know how you do! :)
